### PR TITLE
Multi domain support

### PR DESCRIPTION
**Pull request by jbboehr**
This is super preliminary, just wanted to get the code in a place where it can be reviewed. Some notes:

* I opted to go with `cfg.domain` and `cfg.extraDomains` over `cfg.domains` - it simplified the logic in a few places. Notably, in my current implementation, self-signed certs do not support extraDomains.
* ACME supports all domains using a single certificate - as far as I can tell postfix doesn't have an option to use multiple certs. The certificate is placed in `/var/lib/acme/mailserver/` instead of in any domain subfolder. Certs are also issued independently for each domain and placed in the corresponding domain folder. This is so I don't have to re-implement the nginx ACME configurations.
* I changed the domain that the certs were issued to via ACME due to a conflict with my personal webserver. Only `mail.example.com`, not `example.com` will have an nginx virtualhost set up and certs issued.
* Usernames without a domain will have `@${cfg.domain}` appended to them - another reason I chose `cfg.extraDomains` 
* I added a `.editorconfig` (NixIDEA was having trouble with the indentation) let me know if you want me to remove it.

I've only done minimal testing so far, but it appears to be working on my mail server.

### CODE REVIEW

**Comment by jbboehr** [comment]
@r-raymond My solution works with ACME, but do you know if postfix supports multiple certificates?

**Comment by eqyiel** [comment]
Thanks for working on this, I will check it out over the weekend!

**Comment by r-raymond** [comment]
Awesome!

@jbboehr As far as I can tell no. But it really should not matter as it is only important that the actual FQDN of the server is in the certificate.

Let me explain. Assume I have domains `domain1.com` and `domain2.com` and a server. Assume that the FQDN is e.g. `mx.domain3.com`. I.e. DNS of `mx.domain3.com` points to the server and reverseDNS is set up. Then I'll add `mx.domain3.com` to `MX` records of `domain1.com` and `domain2.com`. Now it is only important that the certificate presented to postfix (and dovecot) is valid for `mx.domain3.com`. The other addresses do not have to be signed.

So `ACME` should only be set up for the `FQDN`. The `FQDN` should be retrieved from `hostname`

**Comment by jbboehr** [comment]
@r-raymond The current setup issues a single ACME certificate valid for all domains listed - this seems preferable to me, but I will test the setup you describe.

**Comment by r-raymond** [comment]
I agree there is now harm done like this (assuming that the actual `FQDN` is also in the certificates). However it implies that the user has to set domain forwarding to the mail server from every domain (otherwise the acme challenges fail). Also I'd be very careful to use `mx.domain1.com` in a `MX` field of `domain1.com` (in the setting of my last post) because rDNS is not going to match then. This is a red herring for many other servers. So it could potentially result in a inferior setup if the user is not careful.

**Comment by jbboehr** [comment]
> Now it is only important that the certificate presented to postfix (and dovecot) is valid for mx.domain3.com. The other addresses do not have to be signed.

Are you sure this will work for end users connecting to the server via IMAP/SMTP/IMAPS/SMTPS though?

To clarify, previously ACME would be configured on `example1.com` and `mail.example1.com` (for `{ domain="example1.com"; hostPrefix="mail"; }`). In my current setup, ACME will be configured on `mail.example1.com` and `mail.example2.com` (for `{ domain="example1.com"; extraDomains=[ "example2.com" ]; hostPrefix = "mail"; }`)

**Comment by r-raymond** [comment]
Absolutely. They would simply specify "mx.domain3.com" as the server (or
thunderbird e.g. would figure that out on it's own from the MX record) and log
in with "user1@domain1.com"

**Comment by r-raymond** [comment]
About the clarification, if you use that setup be careful with rDNS. I.e. tools
like https://mxtoolbox.com/diagnostic.aspx should complain that your rDNS is not
setup correctly (assuming example1.com MX points to mail.example1.com and
example2.com to mail.example2.com). Also the SMTP banner will be wrong. Many
pitfalls that can be avoided.

But really, I should write a setup guide for this in the README

**Comment by jbboehr** [comment]
Ok so to confirm, the optimal DNS setup given the config `{ domain="example1.com"; extraDomains=[ "example2.com" ]; hostPrefix = "mail"; }` is: the MX for `example1.com` and `example2.com` point to `mail.example1.com` The A/CNAME for all three domains point to the IP for `example1.com`. ACME only issues a certicate valid for `example1.com` and `mail.example1.com`.

And this will work for users connecting to either `userA@example1.com` or `userB@example2.com` ?

**Comment by r-raymond** [comment]
This will work perfectly if you do not set the A records. Note that those can
point to any other address completely unrelated to the mail server. If you
_also_ want to run a webserver, then indeed you need your certificate to contain
both example1.com and example2.com (assuming that the webserver is supposed to
run on the same host), so that nginx has a valid certificate for both domains.
However that is unrelated to the mail setup.

If we ever want to include some front end (Sogo or so) _then_ it would be
necessary to add all the other entries also. (Again for nginx). But I'd do that
on a "need to have" basis, because someone might be fine to just simply run a
mailserver and that shouldn't be prevented by the setup.

And yes, this will work for both users from example1.com and example2.com

File: mail-server/postfix.nix [hunk 1]
```diff
@@ -38,7 +40,7 @@ let
   valiases_file = builtins.toFile "valias" (lib.concatStringsSep "\n" valiases_postfix);
 
   # vhosts_file :: Path
-  vhosts_file = builtins.toFile "vhosts" cfg.domain;
+  vhosts_file = builtins.toFile "vhosts" (concatStringsSep ", " allDomains);
```
Comment (r-raymond) [inline mail-server/postfix.nix]: Normally separated by newline in a file, but "," seems to be fine according to the spec

**Comment by jbboehr** [comment]
@r-raymond These are the places where a "primary" domain is needed without further changes.
* Self-signed cert - https://github.com/r-raymond/nixos-mailserver/blob/2f7e3a9f0ce314d03e0e6371a0cbf3085ce83cd5/mail-server/systemd.nix#L26
* DKIM cert - https://github.com/r-raymond/nixos-mailserver/blob/2f7e3a9f0ce314d03e0e6371a0cbf3085ce83cd5/mail-server/systemd.nix#L53
* Self-signed cert - https://github.com/r-raymond/nixos-mailserver/blob/ddfb2e621077378f11e45b2b81e6e0e91974d47c/mail-server/common.nix#L27
* Self-signed cert - https://github.com/r-raymond/nixos-mailserver/blob/ddfb2e621077378f11e45b2b81e6e0e91974d47c/mail-server/common.nix#L36
* ACME cert - in our above discussion, you said the cert only needed to be signed for one domain, so I would need to pick one of them. This should be the one the DNS was configured properly for, but there's no way of knowing that without the user telling us.

The other reason I used the primary domain with qualifyUser is that it should be compatible with the current version (for any other users upgrading) and it's less typing to set up aliases. Note that it's not required to use unqualified usernames with the primary domain.

So you think these should all just use the system `config.hostname` which could be a value which is not in our `config.domains`?

**Comment by r-raymond** [comment]
@jbboehr 

As mentioned in the pull request #23 I think we should just allow the user to specify the fqdn so he can set it to whatever he likes. This would then be similar to your special domain, except that it is not a domain that we actually add to the virtual domains of postfix. If the use wants to do so, he can do that, but it is not necessary. I e.g. have my mail-server running at `mx.a.com`, and it serves only `b.com` and `c.com`, not `a.com`.

The self signing certificates should be easy to fix (${hostPrefix}.${domain} -> ${fqdn}) and DKIM is broken anyway, as we need to add all domains of the virtualhost file to the certificate.

The ACME cert should be simply issued for ${fqdn}.

So basically, I vote for renaming `domain` to `fqdn` and removing it from the list of domains.

File: default.nix [hunk 2]
```diff
@@ -28,8 +28,15 @@ in
 
     domain = mkOption {
       type = types.str;
-      example = "example.com";
-      description = "The domain that this mail server serves. So far only one domain is supported";
+      example = "[ example.com ]";
+      description = "The primary domain that this mail server serves.";
+    };
+
+    extraDomains = mkOption {
```
[the remark this answers is not in the document]
Comment (eqyiel) [reply default.nix]: I agree with this!

**Comment by jbboehr** [comment]
@r-raymond Okay I think I've finally understood you. I will commit my changes shortly. Hopefully you didn't spend longer explaining it to me than it would've taken to implement it.

**Comment by r-raymond** [comment]
Awesome! Don't worry, I wasn't sure myself in the beginning how to best implement this.

Thanks a lot for your time!
